File: 7.ReverseInteger/reverse_integer.cpp
class Solution
{
public:
    // 用除法取模来反转数字
    int reverse(int x)
    {
        int result = 0;
        while (x != 0)
        {
            // 个位数
            int pop_num = x % 10;
            x = x / 10;

            // [−2^31,  2^31 − 1]，题目要求int32最大整数。2147483648，则正数个位要小于7，负数个位要大于-8
            // 如果已经超过INT_MAX/10，则加上后面的pop_num位之后肯定超。或者如果目前等于INT_MAX/10，则需要小于INT_MAX % 10的个位数
            if (result > INT_MAX / 10 || (result == INT_MAX / 10 && pop_num > INT_MAX % 10))
                return 0;
            //负数的情况
            if (result < INT_MIN / 10 || (result == INT_MIN / 10 && pop_num < INT_MIN % 10))
                return 0;
            result = result * 10 + pop_num;
        }
        return result;
    }
};

File: 7.ReverseInteger/reverse_integer.py
class Solution(object):
    def reverse(self, x):
        """
        :type x: int
        :rtype: int
        """

        is_small_zero = False
        if x < 0:
            num_list = list(str(x))[1:]
            is_small_zero = True
        else:
            num_list = list(str(x))

        num_list.reverse()
        if is_small_zero:
            result = int("".join(num_list)) * -1
        else:
            result = int("".join(num_list))

        import math
        if result < math.pow(2,31) * -1  or result > math.pow(2,31)-1:
            return 0
        else:
            return result

if __name__ == "__main__":
    s = Solution()
    x = -1234
    print(s.reverse(x))

